packages/datadog-shimmer/src/shimmer.js

```diff
@@ -1,85 +1,88 @@
 'use strict'
 
+const skipMethods = new Set([
+  'caller',
+  'arguments',
+  'name',
+  'length'
+])
+
 function copyProperties (original, wrapped) {
-  // TODO getPrototypeOf is not fast. Should we instead do this in specific
-  // instrumentations where needed?
-  const proto = Object.getPrototypeOf(original)
-  if (proto !== Function.prototype) {
+  if (original.constructor !== wrapped.constructor) {
+    const proto = Object.getPrototypeOf(original)
     Object.setPrototypeOf(wrapped, proto)
   }
-  const props = Object.getOwnPropertyDescriptors(original)
-  const keys = Reflect.ownKeys(props)
-
-  for (const key of keys) {
-    try {
-      Object.defineProperty(wrapped, key, props[key])
-    } catch (e) {
-      // TODO: figure out how to handle this without a try/catch
+
+  const ownKeys = Reflect.ownKeys(original)
+  if (original.length !== wrapped.length) {
+    Object.defineProperty(wrapped, 'length', { value: original.length, configurable: true })
+  }
+  if (original.name !== wrapped.name) {
+    Object.defineProperty(wrapped, 'name', { value: original.name, configurable: true })
+  }
+  if (ownKeys.length !== 2) {
+    for (const key of ownKeys) {
+      if (skipMethods.has(key)) continue
+      const descriptor = Object.getOwnPropertyDescriptor(original, key)
+      if (descriptor.writable || descriptor.configurable || !Object.hasOwn(wrapped, key)) {
```

Comment:
This line makes the assumption that any property name that exists both on the wrapped and the original has roughly the same descriptor, modulo the value. This is because configurability and writability are checked on the original, but not on the wrapped. If we're already making that assumption, maybe it's safe to make a fast path here and just set the value directly if it's writable?

Or maybe that's a rare enough case that it's not worth bothering with here. WDYT?

Reply:
I am not sure I follow the comment.

The assumption is indeed that any property on a method that is either writable or configurable on one side would also be that on the other side, if it exists. If it does not exist, it does not matter, than we have to set it anyway.
I always need the descriptor of the original while this prevents looking up the descriptor on the wrapper.